[Bug] fails to build rust's static library

### Xmake Version

xmake v3.0.0+HEAD.0db4fe610

### Operating System Version and Architecture

Windows 11 24H2

### Describe Bug

I tested it on an xmake rust's static project template but on compile it errors with `can't find crate for ``foo``;`.

I'm currently on rust:

```console
> rustc --version                                                                                                                                                          
rustc 1.88.0 (6b00bc388 2025-06-23)
```

### Expected Behavior

It should be able to compile and run.

### Project Configuration

```lua
target("foo")
    set_kind("static")
    add_files("src/foo.rs")

target("testing_demo")
    set_kind("binary")
    add_deps("foo")
    add_files("src/main.rs")
```

### Additional Information and Error Logs

```console
> xmake create -l rust -t static -P testing
create testing ...
  [+]: src\foo.rs
  [+]: src\main.rs
  [+]: xmake.lua
  [+]: .gitignore
create ok!

> cd testing
> xmake f -cvD
checking for platform ... windows                                                                                                                                          
checking for architecture ... x64                                                                                                                                          
checking for cl.exe ... D:\Dev\VisualStudio\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.43.34808\bin\HostX64\x64\cl.exe
checking for Microsoft Visual Studio (x64) version ... 2022
checking for Microsoft C/C++ Compiler (x64) version ... 19.43.34810
checkinfo: cannot runv(nim.exe --version), No such file or directory
checking for nim ... no
checkinfo: cannot runv(nim.exe --version), No such file or directory
checking for nim ... no
configure
{
    builddir = build
    host = windows
    mode = release
    ndk_stdcxx = true
    kind = static
    plat = windows
    clean = true
    ccache = true
    arch = x64
    vs = 2022
}
```

on which it errors with `can't find crate for ``foo``;`.

```console
> xmake b -vD
checking for rustc ... D:\Dev\rust\.cargo\bin\rustc.exe                                                                                                                    
checking for the rust compiler (rc) ... rustc.exe                                                                                                                          
[ 31%]: linking.release foo.rlib
checking for rustc ... D:\Dev\rust\.cargo\bin\rustc.exe
checking for the rust static library archiver (rcar) ... rustc.exe
D:\Dev\rust\.cargo\bin\rustc.exe --target=x86_64-pc-windows-msvc --edition 2018 --crate-type=lib -o build\windows\x64\release\foo.rlib src\foo.rs
[ 79%]: linking.release testing_demo.exe
checking for rustc ... D:\Dev\rust\.cargo\bin\rustc.exe
checking for the rust linker (rcld) ... rustc.exe
D:\Dev\rust\.cargo\bin\rustc.exe --target=x86_64-pc-windows-msvc --edition 2018 -Lbuild\windows\x64\release --crate-type=bin -o build\windows\x64\release\testing_demo.exe src\main.rs
error: @programdir\core\main.lua:329: @programdir\actions\build\main.lua:146: @programdir\modules\async\runjobs.lua:331: @programdir\rules\rust\build\target.lua:67: @programdir\core\sandbox\modules\os.lua:273: error[E0463]: can't find crate for `foo`
 --> src\main.rs:1:1
  |
1 | extern crate foo;
  | ^^^^^^^^^^^^^^^^^ can't find crate

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0463`.

stack traceback:
    [C]: in function 'error'
    [@programdir\core\base\os.lua:1075]:
    [@programdir\core\sandbox\modules\os.lua:273]: in function 'runv'
    [@programdir\modules\core\tools\rustc.lua:129]:
    [C]: in function 'xpcall'
    [@programdir\core\base\utils.lua:246]:
    [@programdir\rules\rust\build\target.lua:67]: in function 'build_sourcefiles'
    [@programdir\rules\rust\build\target.lua:83]:
    [@programdir\modules\private\action\build\target.lua:172]: in function 'jobfunc'
    [@programdir\modules\async\runjobs.lua:247]:
    [C]: in function 'xpcall'
    [@programdir\core\base\utils.lua:246]: in function 'trycall'
    [@programdir\core\sandbox\modules\try.lua:117]: in function 'try'
    [@programdir\modules\async\runjobs.lua:230]: in function 'cotask'
    [@programdir\core\base\scheduler.lua:406]:

stack traceback:
        [C]: in function 'error'
        @programdir\core\base\os.lua:1075: in function 'os.raiselevel'
        (...tail calls...)
        @programdir\core\main.lua:329: in upvalue 'cotask'
        @programdir\core\base\scheduler.lua:406: in function <@programdir\core\base\scheduler.lua:399>
```

--- a/xmake/rules/rust/xmake.lua
+++ b/xmake/rules/rust/xmake.lua
@@ -105,10 +105,17 @@ rule("rust.build")
         assert(rc, "rustc not found. Failed to add libstd in env")
         local outdata, errdata = os.iorunv(rc.program, {"--print=sysroot"})
         assert(not errdata or errdata == "", "failed to find rust sysroot:\n" .. errdata)
-        local libstd = path.join(outdata:trim(), "bin")
-        target:add("runenvs", "PATH", libstd)
-        target:add("runenvs", "LD_LIBRARY_PATH", libstd)
-        target:add("runenvs", "DYLD_LIBRARY_PATH", libstd)
+        local rustc_sysroot = outdata:trim()
+        if target:is_plat("windows") then
+            local libstd = path.join(rustc_sysroot, "bin")
+            target:add("runenvs", "PATH", libstd)
+        elseif target:is_plat("macosx") then
+            local libstd = path.join(rustc_sysroot, "lib")
+            target:add("runenvs", "DYLD_LIBRARY_PATH", libstd)
+        else
+            local libstd = path.join(rustc_sysroot, "lib")
+            target:add("runenvs", "LD_LIBRARY_PATH", libstd)
+        end
     end)
 
 rule("rust")

--- a/xmake/rules/rust/xmake.lua
+++ b/xmake/rules/rust/xmake.lua
@@ -97,16 +97,18 @@ rule("rust.build")
     end)
     on_build("build.target")
 
-    before_run(function (target)
+    on_config(function (target)
         import("lib.detect.find_tool")
 
-        -- Make sure libstd shared library is available
+        -- Ensure libstd shared library is available when running
         local rc = find_tool("rustc")
         assert(rc, "rustc not found. Failed to add libstd in env")
         local outdata, errdata = os.iorunv(rc.program, {"--print=sysroot"})
         assert(not errdata or errdata == "", "failed to find rust sysroot:\n" .. errdata)
         local libstd = path.join(outdata:trim(), "bin")
-        os.addenvs({PATH = libstd, LD_LIBRARY_PATH = libstd, DYLD_LIBRARY_PATH = libstd})
+        target:add("runenvs", "PATH", libstd)
+        target:add("runenvs", "LD_LIBRARY_PATH", libstd)
+        target:add("runenvs", "DYLD_LIBRARY_PATH", libstd)
     end)
 
 rule("rust")

--- a/xmake/rules/rust/xmake.lua
+++ b/xmake/rules/rust/xmake.lua
@@ -32,6 +32,9 @@ rule("rust.cxxbridge")
 rule("rust.build")
     set_sourcekinds("rc")
     on_load(function (target)
+        -- set cratename
+        local cratename = target:values("rust.cratename") or target:name()
+        target:set("basename", cratename)
         -- set cratetype
         local cratetype = target:values("rust.cratetype")
         if cratetype == "staticlib" then
@@ -43,19 +46,32 @@ rule("rust.build")
             target:add("shflags", "--crate-type=cdylib", {force = true})
             target:add("shflags", "-C prefer-dynamic", {force = true})
         elseif target:is_static() then
+            cratetype = "lib"
+            target:set("values", "rust.cratetype", cratetype)
+            if is_plat("windows") then
+                target:set("prefixname", "lib")
+            end
             target:set("extension", ".rlib")
-            target:add("arflags", "--crate-type=lib", {force = true})
+            target:add("arflags", "--crate-type=" .. cratetype, {force = true})
             target:data_set("inherit.links.deplink", false)
         elseif target:is_shared() then
-            target:add("shflags", "--crate-type=dylib", {force = true})
+            cratetype = "dylib"
+            target:set("values", "rust.cratetype", cratetype)
+            target:add("shflags", "--crate-type=" .. cratetype, {force = true})
             -- fix cannot satisfy dependencies so `std` only shows up once
             -- https://github.com/rust-lang/rust/issues/19680
             --
             -- but it will link dynamic @rpath/libstd-xxx.dylib,
             -- so we can no longer modify and set other rpath paths
             target:add("shflags", "-C prefer-dynamic", {force = true})
+            -- don't add links and instead use --extern CRATE[=PATH]
+            -- also, it'll automatically link from linkdirs (-L) even if no --extern
+            -- https://github.com/xmake-io/xmake/issues/6604
+            target:data_set("inherit.links.deplink", false)
         elseif target:is_binary() then
-            target:add("ldflags", "--crate-type=bin", {force = true})
+            cratetype = "bin"
+            target:set("values", "rust.cratetype", cratetype)
+            target:add("ldflags", "--crate-type=" .. cratetype, {force = true})
         end
 
         -- set edition

--- a/xmake/rules/rust/xmake.lua
+++ b/xmake/rules/rust/xmake.lua
@@ -97,6 +97,18 @@ rule("rust.build")
     end)
     on_build("build.target")
 
+    before_run(function (target)
+        import("lib.detect.find_tool")
+
+        -- Make sure libstd shared library is available
+        local rc = find_tool("rustc")
+        assert(rc, "rustc not found. Failed to add libstd in env")
+        local outdata, errdata = os.iorunv(rc.program, {"--print=sysroot"})
+        assert(not errdata or errdata == "", "failed to find rust sysroot:\n" .. errdata)
+        local libstd = path.join(outdata:trim(), "bin")
+        os.addenvs({PATH = libstd, LD_LIBRARY_PATH = libstd, DYLD_LIBRARY_PATH = libstd})
+    end)
+
 rule("rust")
     add_deps("rust.build")
     add_deps("utils.inherit.links")

--- a/xmake/rules/utils/inherit_links/inherit_links.lua
+++ b/xmake/rules/utils/inherit_links/inherit_links.lua
@@ -96,9 +96,9 @@ function main(target)
             local cratetype = target:values("rust.cratetype")
             -- only bin, lib, rlib, dylib can make use of --extern CRATE[=PATH]
             if cratetype ~= "staticlib" and cratetype ~= "cdylib" then
-                local cratename = target:values("rust.cratename") or target:name()
-                local extern_crate_opt = string.format('--extern %s=%s', cratename, target:targetfile())
-                _add_export_value(target, "rcldflags", extern_crate_opt)
+                local extern_crate_opt = rustc:nf_framework(target:targetfile())
+                local extern_crate = table.concat(extern_crate_opt, " ")
+                _add_export_value(target, "rcflags", extern_crate)
             end
         end
     end

--- a/xmake/rules/utils/inherit_links/inherit_links.lua
+++ b/xmake/rules/utils/inherit_links/inherit_links.lua
@@ -91,6 +91,16 @@ function main(target)
             -- we need to add includedirs to support import modules for golang
             _add_export_value(target, "includedirs", path.directory(targetfile))
         end
+
+        if target:rule("rust") then
+            local cratetype = target:values("rust.cratetype")
+            -- only bin, lib, rlib, dylib can make use of --extern CRATE[=PATH]
+            if cratetype ~= "staticlib" and cratetype ~= "cdylib" then
+                local cratename = target:values("rust.cratename") or target:name()
+                local extern_crate_opt = string.format('--extern %s=%s', cratename, target:targetfile())
+                _add_export_value(target, "rcldflags", extern_crate_opt)
+            end
+        end
     end
 
     -- we export all links and linkdirs in self/packages/options to the parent target by default

--- a/xmake/rules/utils/inherit_links/inherit_links.lua
+++ b/xmake/rules/utils/inherit_links/inherit_links.lua
@@ -96,9 +96,7 @@ function main(target)
             local cratetype = target:values("rust.cratetype")
             -- only bin, lib, rlib, dylib can make use of --extern CRATE[=PATH]
             if cratetype ~= "staticlib" and cratetype ~= "cdylib" then
-                local extern_crate_opt = rustc:nf_framework(target:targetfile())
-                local extern_crate = table.concat(extern_crate_opt, " ")
-                _add_export_value(target, "rcflags", extern_crate)
+                _add_export_value(target, "frameworks", target:targetfile())
             end
         end
     end

--- a/xmake/templates/rust/shared/project/src/foo.rs
+++ b/xmake/templates/rust/shared/project/src/foo.rs
@@ -1,0 +1,4 @@
+pub fn add(a: i32, b: i32) -> i32 {
+    return a + b;
+}
+

--- a/xmake/templates/rust/shared/project/src/main.rs
+++ b/xmake/templates/rust/shared/project/src/main.rs
@@ -1,0 +1,6 @@
+extern crate foo;
+
+fn main() {
+    println!("hello xmake!");
+    println!("add: {}", foo::add(1, 1));
+}

--- a/xmake/templates/rust/shared/project/xmake.lua
+++ b/xmake/templates/rust/shared/project/xmake.lua
@@ -7,12 +7,4 @@ target("${TARGETNAME}_demo")
     add_deps("foo")
     add_files("src/main.rs")
 
-    --- Make sure libstd shared library is available
-    before_run(function (target)
-        local outdata, errdata = os.iorun("rustc --print=sysroot")
-        assert(not errdata or errdata == "", "failed to find rust sysroot:\n" .. errdata)
-        local libstd = path.join(outdata:trim(), "bin")
-        os.addenvs({PATH = libstd, LD_LIBRARY_PATH = libstd, DYLD_LIBRARY_PATH = libstd})
-    end)
-
 ${FAQ}

--- a/xmake/templates/rust/shared/project/xmake.lua
+++ b/xmake/templates/rust/shared/project/xmake.lua
@@ -1,0 +1,18 @@
+target("foo")
+    set_kind("shared")
+    add_files("src/foo.rs")
+
+target("${TARGETNAME}_demo")
+    set_kind("binary")
+    add_deps("foo")
+    add_files("src/main.rs")
+
+    --- Make sure libstd shared library is available
+    before_run(function (target)
+        local outdata, errdata = os.iorun("rustc --print=sysroot")
+        assert(not errdata or errdata == "", "failed to find rust sysroot:\n" .. errdata)
+        local libstd = path.join(outdata:trim(), "bin")
+        os.addenvs({PATH = libstd, LD_LIBRARY_PATH = libstd, DYLD_LIBRARY_PATH = libstd})
+    end)
+
+${FAQ}

--- a/xmake/templates/rust/shared/template.lua
+++ b/xmake/templates/rust/shared/template.lua
@@ -1,0 +1,2 @@
+template("shared")
+    add_configfiles("xmake.lua")

